src/include/simeng/memory/Mem.hh

```diff
@@ -1,74 +1,120 @@
+enum DataPacketType {
+  /** This type signifies a read request to memory. */
+  READ_REQUEST,
 
-/* Response to a read packed. */
-struct ReadRespPacket : public DataPacket {
-  uint64_t reqId;
-  size_t bytesRead;
-  char* data;
-  ReadRespPacket(uint64_t req_id, size_t bytes_read, char* dt);
-};
+  /** This type signifies a write request to memory.*/
+  WRITE_REQUEST,
 
-/* ReadPacket represents a read access to data packet. */
-struct ReadPacket : public DataPacket {
-  uint64_t address;
-  size_t size;
+  /** This type signifies the response to a read request. */
+  READ_RESPONSE,
 
-  ReadPacket(uint64_t addr, size_t sz);
-  ReadRespPacket* makeResponse(uint64_t bytesRead, char* data);
-};
+  /** This type signifies the response to a write request. */
+  WRITE_RESPONSE,
 
-/* Response to a write packed. */
-struct WriteRespPacket : public DataPacket {
-  uint64_t reqId;
-  size_t bytesWritten;
-  WriteRespPacket(uint64_t req_id, size_t bytes_written);
+  /** This type signifies a faulty or empty request of no type. */
+  NONE
 };
 
-/* WritePacket represents a write access to a data packet. */
-struct WritePacket : public DataPacket {
-  uint64_t address;
-  size_t size;
-  const char* data;
+/* A data packet struct contains information which enables access to the
+ * simulation memory. */
+struct DataPacket {
+  /** The address at which the memory is to be accessed. */
+  uint64_t address_ = 0;
+
+  /** The size of the memory operation to be performed. */
+  uint64_t size_ = 0;
+
+  /** The type of a DataPacket. */
+  DataPacketType type_ = NONE;
+
+  /** The id of a DataPacket, this is set by the memory interface. For write
+   * requests it is 0 as the memory interface doesn't specify an id for write
+   * requests. */
+  uint64_t id_ = 0;
+
+  /** The data carried by a DataPacket, this is used to deliver and recieve data
+   * to/from the simulation memory. */
+  std::vector<char> data_;
+
+  /** This bool signifies if the memory operation specified by the DataPacket
+   * resulted in a fault. */
+  bool inFault_ = false;
+
+  /** Default constructor for the DataPacket. */
+  DataPacket() {}
+
+  /** Constructor to create a faulty DataPacket. */
+  DataPacket(bool fault) : inFault_(true) {}
 
-  WritePacket(uint64_t addr, size_t sz, const char* dt);
-  WriteRespPacket* makeResponse(uint64_t bytesReturned);
+  /** Constructor for DataPackets which do not hold any data. */
+  DataPacket(uint64_t address, uint64_t size, DataPacketType type,
+             uint64_t reqId, bool fault = false);
+
+  /** Constructor for DataPackers which hold data. */
+  DataPacket(uint64_t address, uint64_t size, DataPacketType type,
+             uint64_t reqId, std::vector<char> data, bool fault = false);
+
+  /** Copy constructor for DataPacket. */
+  DataPacket(const DataPacket& packet);
+
+  /** Move constructor for DataPacket to enable copy elision whenever it is
+   * possible. */
+  DataPacket(DataPacket&& packet);
```

Comment:
It is worth adding to both comments that these functions have been added to enable copy & move elision.

Also, would this work if we explicitly define the copy/move constructors but set them to default? i.e. `DataPacket(const DataPacket& packet) = 0;`

Reply:
Doesn't `= 0` mean pure virtual which means a derived class has to re-define it?

Reply:
Yes you're correct, sorry, I meant `... = default;`

Reply:
done